### PR TITLE
Ensure successful unmarshaling of service type in Manifest struct

### PR DESCRIPTION
**Pull request by ckganesan**
This PR is raised to address and resolve the issue. [#392](https://github.com/cloudfoundry-community/go-cfclient/issues/392)

### CODE REVIEW

File: operation/manifest.go [hunk 1]
```diff
@@ -81,6 +81,36 @@ type AppManifestService struct {
...
 
+func (ams *AppManifestService) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var raw interface{}
+	if err := unmarshal(&raw); err != nil {
+		return err
+	}
+	switch v := raw.(type) {
+	case string:
+		ams.Name = v
+	case map[interface{}]interface{}:
```
Comment (Gourab1998) [inline operation/manifest.go:92]: When we try unmarshaling from raw yaml type directly it needs map[string{}]interface{} , hence not getting into this case due to map[interface{}]interface{} type check
Please change to map[string]interface{}

image

File: operation/manifest.go [hunk 2]
```diff
@@ -81,6 +81,36 @@ type AppManifestService struct {
...
+		for key, value := range v {
+			switch key {
+			case "name":
+				ams.Name = value.(string)
+			case "binding_name":
+				ams.BindingName = value.(string)
+			case "parameters":
+				if params, ok := value.(map[interface{}]interface{}); ok {
```
Comment (Gourab1998) [inline operation/manifest.go:100]: Same here needs map[string]interface{}

File: operation/manifest.go [hunk 3]
```diff
@@ -81,6 +83,49 @@ type AppManifestService struct {
...
 
+func (ams *AppManifestService) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var raw interface{}
+	if err := unmarshal(&raw); err != nil {
+		return err
+	}
+	switch v := raw.(type) {
+	case string:
+		ams.Name = v
+	case map[interface{}]interface{}:
```
Comment (Gourab1998) [inline operation/manifest.go:94]: It can never be map[interface{}]interface{} type. 
Please let me know if I am missing something if not do remove it.

Comment (ckganesan) [reply operation/manifest.go:94]: It seems like you are utilizing the yaml.v3 library, whereas yaml.v2 requires the use of map[interface{}]interface{}.

Comment (Gourab1998) [reply operation/manifest.go:94]: Thanks for the info!

**Comment by sneal** [comment]
Can we add a test that validates we can unmarshal _both_ representations?

**Comment by Gourab1998** [comment]
@ckganesan need this PR merged ASAP, if you have capacity crunch I can contribute testcases for this, just let me know.

**Comment by ckganesan** [comment]
I've included test cases for manifest YAML unmarshalling.

**Comment by sneal** [comment]
Looks like we just need a test case for unmarshalling the services attribute without a name attribute
```yaml
  services:
  - xsuaa-tf
  ```

File: operation/manifest_test.go [hunk 4]
```diff
@@ -5,8 +5,67 @@ import (
...
+    env:
+      SPRING_CLOUD_PROFILE: dev
+    routes:
+      - route: spring-music-egregious-porcupine-oa.apps.example.org
+    services:
+      - name: my-sql
+        binding_name: mysql
+        parameters:
+          name: mysql
+      - oauth2
```
Comment (ckganesan) [inline operation/manifest_test.go:57]: I have already included the mentioned cases as well.

Comment (sneal) [reply operation/manifest_test.go:57]: I missed that oauth2 service, thanks for adding that!

**Pull request merged**